### PR TITLE
Update signup goals page skip to dashboard set site_intent to build

### PR DESCRIPTION
**Pull request by agrullon95**
#### Testing/Review instructions

Test: short
Review: short

#### Proposed Changes

Currently, users that click the `Skip to dashboard` link on the top right-hand corner of the Goals onboarding step will be navigated to the home screen with the following site options:

`site_intent` = ""
`launchpad_screen` = false

The changes in this PR will set the `site_intent` site option to `build` if the user clicks `Skip to dashboard` on the Goals page. This will also enable the launchpad and set the `launchpad_screen` site option to `full` before the user is navigated to the home screen.

P2 post to give heads up: paYE8P-2yP-p2

Goals step:
image

#### Testing Instructions

* Checkout this branch
* Navigate to the onboarding flow (Click `< Switch Site` on the Calypso sidebar and then click `Add a new site`)
* Create your free site and you should land on the Goals step.
* Do not select a goal and click `Skip to dashboard`. The `site_intent` and `launchpad_screen` values should be updated. The home screen should redirect you to the launchpad screen (There may be a bug that requires multiple refreshes before you redirected - )
* Create a new site and test another goal such as `Write & Publish` to verify the correct `site_intent` value is set

#### Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Related to #72870

### CODE REVIEW

**Comment by Sharpirate** [comment]
### Testing
Go to `http://calypso.localhost:3000/start`

**Comment by sharpirate** [comment]
### Testing
1. Go to `http://calypso.localhost:3000/start`
2. On the Goals page click on "Skip to dashboard"
3. On My Home refresh we are redirected to Launchpad which indicates the site options were updated successfully.
5. Tested with the `Write & Publish` flow continues to work when selecting Goals.

**Review by sharpirate — APPROVED** [review]

**Comment by jeyip** [comment]
## Testing
### Requirements
- [x] Do not select a goal and click Skip to dashboard. The site_intent and launchpad_screen values should be updated. The home screen should redirect you to the launchpad screen
- [x] Create a new site and test another goal such as Write & Publish to verify the correct site_intent value is set
- [x] Could complete Launchpad tasks and launch site as expected
- [x] Launchpad iframe overlay, url copy to clipboard button, customize pill, and go to admin button continued to work as expected 

### Browsers
- [x] Chrome

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (Addison-Stavlo) [inline client/landing/stepper/declarative-flow/site-setup-flow.ts]: Note this isnt specifically tied to this dashboard button on the goals step, so im curious if there are any other areas this change could potentially bleed into?

That said, testing around and escaping out of other flows early I see the intent the same as expected. So it seems to be looking good in my testing.

**Review by Addison-Stavlo — APPROVED** [review]
This tests well on my end, minus the other issue tracked elsewhere that shouldnt block this PR. https://github.com/Automattic/wp-calypso/issues/72872

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (jeyip) [inline client/landing/stepper/declarative-flow/site-setup-flow.ts]: Since we're setting a default value for _any_ flow with some falsey value for `intent`, would we inadvertently enable launchpad for situations where it's undesired?

I can't think of anything off the top of my head, but I suppose, if we wanted to be safer, we could only set default values for general onboarding

**Review by jeyip — COMMENTED** [review]
I left one comment. +1 from me when we decide how we want to approach the concern

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (agrullon95) [inline client/landing/stepper/declarative-flow/site-setup-flow.ts:550]: Sets site_intent to `build` in redux store if user clicks "Skip to dashboard"

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (agrullon95) [inline client/landing/stepper/declarative-flow/site-setup-flow.ts:221]: Added `skipIntent` flag to prevent the site intent from being deleted before its used in the setPendingAction  function above.

File: packages/data-stores/src/onboard/reducer.ts
Comment (agrullon95) [inline packages/data-stores/src/onboard/reducer.ts:310]: If we reset the onboard store and include the `skipIntent` flag, then we don't want to delete the site_intent state value.

**Comment by sharpirate** [comment]
✅ Still tests well after the changes

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (Addison-Stavlo) [inline client/landing/stepper/declarative-flow/site-setup-flow.ts:185]: why cant we use the regular `intent` here (and below)? Is there a reason we need to have both:

```
		const intent = useSelect( ( select ) => select( ONBOARD_STORE ).getIntent() );
		const { getIntent } = useSelect( ( select ) => select( ONBOARD_STORE ) );
```

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (Addison-Stavlo) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:221]: I dont see this reset being called in many other places. So it seems like we are adding `skipIntent` to the main case here?

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (Addison-Stavlo) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:550]: Im not understanding the need for the other changes, but this seems like the main thing we need here?

Comment (agrullon95) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:550]: Yes this is the main change, but I needed to add the other changes in order to get the latest site_intent value.

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (agrullon95) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:185]: Im not able to get the correct value from using the existing `intent` variable. It appears as an empty string.

**Review by Addison-Stavlo — COMMENTED** [review]
Behavior tests well on my end 👍 

To be safe and communicative here, would you mind writing a P2 about this and crossposting for dotcom and related teams? essentially that we are making this "skip" from this goals page default to `build`, similar to how "other" flow defaults to `build`? Mostly as a heads up, but also to double check for any concerns with this we may have missed?

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (agrullon95) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:221]: Yes im adding it here to ensure Im still able to retrieve the site intent value from redux inside the `exitFlow` pendingActions code.

Without adding this and modifying redux, I noticed that site intent value was always an empty string when I used the `getIntent` selector.

File: client/landing/stepper/declarative-flow/site-setup-flow.ts
Comment (edanzer) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:185]: I was going to ask the same question. Any thoughts on why? I assume it may a JS timing issue related to how and when pendingAction is invoked? Like in one case, you're getting the value of intent NOW as you set up the function (essentially acts as a closure and stores the current value), and in the other, your fetching site intent later, when the method itself is fired. 

I will say it seems odd in this file to be fetching intent here twice via two different methods, in two different places.

Comment (edanzer) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:185]: I think one other thing I'm puzzle by here is that before this PR, we WERE using the first/original intent value. Does that mean that code was not working as expected?

Comment (edanzer) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:185]: I'm a bit concerned about whether this change (since it clearly does change the behavior and possible value of intent) will have unexpected impact on other existing flows beyond what we're wanting to do just for ours.

Comment (edanzer) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:185]: Reviewed this @agrullon95 and just did some extensive testing of other scenarios and flows to confirm I'm not seeing any issues.

**Review by edanzer — APPROVED** [review]
This tests as expected. I have some concerns that these code changes could plausibly impact how intent is set in a range of circumstances. I reviewed with @agrullon95 and I tested both the intended scenario (Skip to Dashboard) but also a wide mix of other flows and scenarios, and everything tests as expected. 

In another comment @Addison-Stavlo mentioned maybe posting about this. Not sure if it's worth flagging one of the people who worked on the site-setup-flow file to confirm they're good with it too? 

All that said, given all the checks and testing, I'm also approving now.

**Comment by agrullon95** [comment]
@edanzer - I added a link to the P2 in the description where I gave other teams a heads-up about these changes

**Pull request merged**
